js/customize-snapshot.js

```diff
@@ -0,0 +1,166 @@
+/* global jQuery, _customizeWidgetsPlusCustomizeSnapshot, JSON */
+/* exported customizeSnapshot */
+var customizeSnapshot = ( function( $ ) {
+
+	var self = {},
+		api = wp.customize,
+		uuid = _customizeWidgetsPlusCustomizeSnapshot.uuid,
+		is_preview = _customizeWidgetsPlusCustomizeSnapshot.is_preview,
+		dialog, form;
+
+	/**
+	 * Inject the functionality.
+	 */
+	self.init = function() {
+		api.bind( 'ready', function() {
+			self.addButton();
+			self.addFormDialog();
+			self.addShareDialog();
+
+			dialog = $( '#snapshot-dialog-form' ).dialog( {
+				autoOpen: false,
+				modal: true,
+				buttons: {
+					Save: {
+						text: _customizeWidgetsPlusCustomizeSnapshot.i18n.saveButton,
+						click: self.doAjax
+					},
+					Cancel: {
+						text: _customizeWidgetsPlusCustomizeSnapshot.i18n.cancelButton,
+						click: function() {
+							dialog.dialog( 'close' );
+						}
+					}
+				},
+				close: function() {
+					form[ 0 ].reset();
+				}
+			} );
+
+			form = dialog.find( 'form' ).on( 'submit', function( event ) {
+				event.preventDefault();
+				self.doAjax();
+			} );
+
+			$( '#customize-snapshot' ).on( 'click', function( event ) {
+				event.preventDefault();
+				dialog.dialog( 'open' );
+				dialog.find( 'form input[name=scope]' ).blur();
+				dialog.next( '.ui-dialog-buttonpane' ).find( 'button' ).blur();
+			} );
+		} );
+	};
+
+	/**
+	 * Create the share button.
+	 */
+	self.addButton = function() {
+		var $header = $( '#customize-header-actions' ),
+			$button = '<button id="customize-snapshot" class="dashicons dashicons-share"><span class="screen-reader-text">' + _customizeWidgetsPlusCustomizeSnapshot.i18n.shareButton + '</span></button>';
+
+		if ( $header.length ) {
+			$header.addClass( 'customize-snapshots' ).append( $button );
+		}
+	};
+
+	/**
+	 * Create the dialog form.
+	 */
+	self.addFormDialog = function() {
+		var html = '<div id="snapshot-dialog-form" title="' + _customizeWidgetsPlusCustomizeSnapshot.i18n.formTitle + '">' +
+			'<form>' +
+				'<fieldset>';
+				if ( is_preview ) {
+					html += '<p>' + _customizeWidgetsPlusCustomizeSnapshot.i18n.updateMsg + '</p>' +
+					'<input type="hidden" value="' + _customizeWidgetsPlusCustomizeSnapshot.scope + '" name="scope"></label>';
+				} else {
+					html += '<label for="type-0"><input id="type-0" type="radio" checked="checked" value="dirty" name="scope">' + _customizeWidgetsPlusCustomizeSnapshot.i18n.dirtyLabel + '</label><br>' +
+					'<label for="type-1"><input id="type-1" type="radio" value="full" name="scope">' + _customizeWidgetsPlusCustomizeSnapshot.i18n.fullLabel + '</label><br>';
+				}
+				html += '<input type="submit" tabindex="-1" style="position:absolute; top:-5000px">' +
+				'</fieldset>' +
+			'</form>' +
+		'</div>';
```

Comment:
We should use Underscore templates here. String concatenation of HTML will be an XSS red flag from VIP.